#Sprint 11 Add database

kasisaki opened the pull request.

kasisaki committed:
#sprint11 added scheme tables creation

kasisaki committed:
#sprint11
changed schema.sql order
added enums
changed film model

kasisaki committed:
added
userCreate
findUser
FindDuplicate
findAllUsers

kasisaki committed:
usersDone

kasisaki committed:
findFilms
update
getAllFilms

kasisaki committed:
before changing Genre[] to Set<Genre>

kasisaki committed:
final version

kasisaki committed:
final version

kasisaki committed:
final version

kasisaki committed:
final version

kasisaki committed:
final version

krylov-sergey-sergeevich commented on pom.xml:
```diff
@@ -55,6 +55,18 @@
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-data-jdbc</artifactId>
+			<version>3.0.2</version>
```
А зачем версия указана мы же ее автоматически определяем из спринг родителя

krylov-sergey-sergeevich reviewed (changes requested):
Привет, я в дороге и 7 буду дома поэтому давай первую итерацию сделаем поверхностной и далее будем глубже.

Что бросается в глаза:
1) дубли в билдерах и запросах по возможности давай сделаем где надо специальные методы и избавимся от дублей
2) на счёт уникальности - есть в sql unique штука поизучай и избавиться от методов лишних получится (хотя я вижу что она используется), так вот мы можем не проверять руками уникальность а верить что она сработает или будет исключение и ловить его

Предлагаю подумать и обсудить этот момент

kasisaki committed:
moved business logic to UserService

kasisaki committed:
before test

kasisaki committed:
usersSolved

kasisaki committed:
Все переделал))) Final try Дубль 2

krylov-sergey-sergeevich commented on src/main/java/ru/yandex/practicum/filmorate/controller/film/GenreController.java:
```diff
@@ -0,0 +1,30 @@
+    private final GenreService genreService;
+
+    @GetMapping
+    public List<Genre> findAll() {
```
getAllGenre

krylov-sergey-sergeevich commented on src/main/java/ru/yandex/practicum/filmorate/controller/film/MpaController.java:
```diff
@@ -0,0 +1,34 @@
+
+    private final MpaService mpaService;
+    @GetMapping
+    public List<Mpa> findAll() {
```
getAllMpa

krylov-sergey-sergeevich commented on src/main/java/ru/yandex/practicum/filmorate/controller/film/MpaController.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<Mpa> findGenre(@PathVariable(required = false) int id) {
```
get

krylov-sergey-sergeevich commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -0,0 +1,116 @@
+    private final MpaService mpaService;
+
+
+    public List<Film> findAll() {
```
getAll

krylov-sergey-sergeevich commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -0,0 +1,116 @@
+        return buildFilmsList(filmStorage.findAll());
+    }
+
+    public Film findFilm(int id) {
```
getFilm

krylov-sergey-sergeevich commented on src/main/java/ru/yandex/practicum/filmorate/service/film/GenreService.java:
```diff
@@ -0,0 +1,59 @@
+public class GenreService {
+    private final GenreStorage genreStorage;
+
+    public List<Genre> findAllGenres() {
```
getAllGenres

krylov-sergey-sergeevich commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -0,0 +1,129 @@
+        throw new ElementNotFoundException(String.format("User %d is not in your friendList", friendId));
+    }
+
+    public List<User> findFriends(int id) {
```
id плохое и не очень понятное название для переменной метода

krylov-sergey-sergeevich commented on src/main/java/ru/yandex/practicum/filmorate/util/enums/GenreEnum.java:
```diff
@@ -0,0 +1,20 @@
+package ru.yandex.practicum.filmorate.util.enums;
+
+public enum GenreEnum {
```
А для чего это?

krylov-sergey-sergeevich commented on src/main/java/ru/yandex/practicum/filmorate/util/enums/MpaEnum.java:
```diff
@@ -0,0 +1,19 @@
+package ru.yandex.practicum.filmorate.util.enums;
+
+public enum MpaEnum {
```
А для чего это?

krylov-sergey-sergeevich commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,54 @@
+DROP ALL OBJECTS DELETE FILES;
```
not null в схеме давай подумаем и добавим где еще можно поставить

krylov-sergey-sergeevich replied:
рекомендую primary key делать как constraint внизу в определение таблицы, но это не обязательно

krylov-sergey-sergeevich reviewed (changes requested):
Привет, в целом неплохо, но надо еще поработать над деталями

- А тесты не добавлялись? По тз просто говорится о интеграционном тестирование
- Так-же в файле README.md должна быть ссылка на файл диаграммы

Прошу учесть и переработать имена методов:
- get - используем там где возвращаем или объект или исключение
- find - где или объект или null

Буду ждать с правками =)

kasisaki committed:
Добавил диаграмму в README.md
Изменил сигнатуру методов (find, get)
переименовал переменные методов
Удалил Enum
Добавил интеграционные тесты
Добавил NOT NULL согласно моделям
перместил constrains вниз

krylov-sergey-sergeevich reviewed (changes requested):
Привет, в целом все неплохо, осталось над тестами чутка поработать.

Тесты у нас могут быть интеграционные или junit или junit+mockmvc. Поэтому давай попробуем реализовать тесты подобного формата.

1) Для контроллеров
```java
@WebMvcTest
@AutoConfigureMockMvc
```

2) Для интеграционных тестов
```java
@SpringBootTest
@AutoConfigureTestDatabase
@RequiredArgsConstructor(onConstructor_ = @Autowired)
@SqlGroup({
		@Sql(scripts = {"/schema.sql"},config = @SqlConfig(encoding = "UTF-8")),
		@Sql(executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD, scripts = "/test-data.sql"),
		@Sql(executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD, scripts = "/drop-test-data.sql"),
})
```

Если что пиши в пачке подскажу.

krylov-sergey-sergeevich commented:
Так-же Order нехорошо использовать, это bad practices

kasisaki committed:
Котроллеры тестируются с MockMvc
Убрал аннотоцию @Order

krylov-sergey-sergeevich reviewed (approved):
Привет - зачет!
